hal/cpu_uid_api.h

```diff
@@ -0,0 +1,64 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_CPU_UID_API_H
+#define MBED_CPU_UID_API_H
+
+#include <stddef.h>
+#include "device.h"
+
+#if DEVICE_CPUUID
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_cpu_uid CPUUID hal functions
+ * @{
+ */
+
+
+/** Get length of CPU UID in bytes
+ * 
+ * @return Number of uid bytes
+ * 
+ * @note
+ *  Shall be used by the driver to get the needed size of the byte buffer.
+ *  Target implementation must return a value higher than 0 or the driver will assert.
+ * 
+ */
+int cpu_uid_get_length(void);
```

Comment:
Each target should define a macro:

```C
MBED_CPU_UID_SIZE
```

When a size is defined by a target, it is known at compile time and can never change, so using accessors for it is redundant.

Reply:
Ok good idea.

But where should such a macro belong?

We could define that in targets.json via the "macros_add" directive ...
But when I search the mbed-os folder for things like stdio standard baud rate, there are several target specific approaches - some define it in PeriphalNames.h, some in PinNames.h

Is there some convention for that kind of stuff?

Reply:
Yes, use config please for this. I noticed you used macros_add, that is also an option but having targets config for this is better. look at other targets that have config defined, its usually help string, name and defined value.

Reply:
Ok I just tried this out ... I created a file mbed_lib.json in targets/TARGET_Silicon_Labs/TARGET_EFM32 with following contents:

{
    "name": "hal",
    "config": {
        "deviceuid-size": {
            "help": "Size of the Device's UID (unit Bytes))",
            "value": 8
        }
    }
}

The generated macro is then called MBED_CONF_HAL_DEVICEUID_SIZE.

I see one problem here: When I do an export with "mbed export -i GCC_ARM", this macro will be defined in the mbed_config.h file ... I thought this file is used to change config easily ... there are things like stdio baud rate and so on ...

But the length of the uid shall under no circumstances be changed by the user.

So maybe config is not really the right place for this?

Reply:
It can be overridden, same is true for macros. they can be added/removed. I was referring to `targets.json`, where you can find config per target (see https://github.com/ARMmbed/mbed-os/blob/master/targets/targets.json#L1105). That belongs to a target, and can be overridden by an app in targets_overrides section. Config header file should not be changed, rather config, it's there for tools to include , so we do not pollute much more `-D` option.

We do not have currently readable configuration only.